---
title: "lab10_lorenestasiuk"
author: "Lorene Stasiuk"
date: "11/2/2020"
output: html_document
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library("vcfR")
# I was able to download this package!!!
```

# Part I: Reading VCF data

```{r, vcf_file_str}
data(vcfR_example)

vcf
```

```{r, meta_region}
strwrap(vcf@meta[1:7])

# The meta region may include long lines that may not be easy to view. In vcfR we’ve created a function to help press this data.

queryMETA(vcf)

#When the function queryMETA() is called with only a vcfR object as a parameter, it attempts to summarize the meta information. Not all of the information is returned.
# This is an attempt to summarize information that may be most useful for comprehension of the file’s contents.

queryMETA(vcf, element = 'DP')
```

```{r, fix_region}
# The function getFIX() will suppress IINFO column by default.
head(getFIX(vcf))

```

```{r, gt_region}
# The gt (genotype) region contains information about each variant for each sample. The values for each variant and each sample are colon delimited. Multiple types of data for each genotype may be stored in this manner. The format of the data is specified by the FORMAT column (column nine).

vcf@gt[1:6, 1:4]
```

```{r, vcfr}
# Using the R package vcfR, we can read VCF format files into memory using the function read.vcfR(). 

vcf <- read.vcfR("Grunwald/pinfsc50_filtered.vcf.gz")

# Once in memory we can use the head() method to summarize the information in the three VCF regions.

head(vcf)

#After we have made any manipulations of the file we can save it as a VCF file with the function write.vcf().

write.vcf(vcf, "myVCFdata_filtered.vcf.gz")
```

## Exercises Part I
```{r, exercise1}
# 1) How would we find more information about read.vcfR()?

read.vcfR

# 2) How would we learn what the acronym “AD” stands for?

queryMETA(vcf, element = 'AD')

# 3) We used the head() function to view the first few lines of fix data. How would we view the last few lines of fix data?

tail(vcf@fix)

# 4) There is a column in the fix portion of the data called QUAL. It is not defined in the meta portion of the data because it is defined in the VCF specification. It stands for ‘quality’. Does QUAL appear useful to us? Why or why not?

plot(vcf)

# 5) How would we query the sample names?

colnames(vcf@gt)
```

# Part II: Analysis of Genome Data

```{r, open_examine_data}
# We’ll read our VCF data into R using the function read.vcfR().

vcf <- read.vcfR("Grunwald/pinfsc50_filtered.vcf.gz")

# Once the file is read in we can validate its contents using the show method which is implemented by executing the object’s name at the prompt.

vcf
```

```{r, convert_to_genlight}
# Here we use the function vcfR2genlight() to convert our vcfR object to a genlight object. This makes our VCF data available to the analyses in adegenet.
library("adegenet")

x <- vcfR2genlight(vcf)

# Call object x

x

# vcfR
gt <- extract.gt(vcf, element = "GT")
gt[c(2,6,18), 1:3]

# genlight
t(as.matrix(x))[c(1,5,17), 1:3]


pop(x) <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
popNames(x)


#Here we’ll set the ploidy of all the samples in the genlight object to the same ploidy.

ploidy(x) <- 2
```

```{r, dist_matrices}
# Let’s create a pairwise genetic distance matrix for individuals or populations (i.e., groups of individuals).

# To summarize, we can create a distance matrix from a genlight object using dist():

x.dist <- dist(x)

# There are also functions to create distance matrices from genlight objects that exist in other packages. The function bitwise.dist() in the package poppr is an example.We can use this function as follows.

library("poppr")

x.dist <- poppr::bitwise.dist(x)

# Note, that the variable x has not yet been specified. Lastly, because you can use as.matrix() on your genlight object, and most distance algorithms can use this matrix as input, you can use this as an intermediate step to create a matrix from your genlight object and pass it to your distance algorithm of choice
```

```{r, chromr_obj}
# In this example we will begin by locating the example data from the pinfsc50 package.

library(vcfR)

# Find the files.
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = "pinfsc50")
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = "pinfsc50")
gff_file <- system.file("extdata", "pinf_sc50.gff", package = "pinfsc50")

# Input the files.
vcf <- read.vcfR(vcf_file, verbose = FALSE)
dna <- ape::read.dna(dna_file, format = "fasta")
gff <- read.table(gff_file, sep="\t", quote="")

# Create a chromR object.
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=TRUE)
```


```{r, chromr_obj2}
# We invoke the object’s ‘show’ method. The show method for chromR objects presents a summary of the object’s contents.

chrom

# There at least two ways to graphically view the chromR object. The first is plot() which plots histograms of some of data summaries.

plot(chrom)

# The read depth here is a sum over all samples. We see a peak that represents the depth where most of our genomes were sequenced at.
```


```{r, chromr_obj3}
# The simplest interpretation here is that QUAL may not be a good parameter to use to judge your variants. The last panel for SNP densities is empty because this data is created during the processing of chromR objects

chromoqc(chrom, dp.alpha = 58)

# Our second plot, called chromo plot, displays the same information as the plot method only it distributes the data along its chomosomal coordinates.
```

```{r, process_chromr}
# Creation and processing of a chromR object has been divided into separate tasks. Creation loads the data into the chromR object and should typically only be required once. Processing the chromR object generates summaries of the data. Some of these summaries will need to be updated as the chromR object is updated.

chrom <- proc.chromR(chrom, verbose = TRUE)

plot(chrom)

chromoqc(chrom, dp.alpha = 66)

# The above data is an example of visualizing raw data that has come from a variant caller and other automated sources.
```

```{r, process_chromr1}
vcf <- read.vcfR("Grunwald/pinfsc50_filtered.vcf.gz", verbose = FALSE)

chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=FALSE)

chrom <- proc.chromR(chrom, verbose = FALSE)

chromoqc(chrom, dp.alpha = 66)
```

```{r, tab_sum}
# When we process a chromR object, two forms of tabular data are created. First, summaries are made on a per variant basis. Second, summaries are made on a per window basis. 

head(chrom@var.info)

# Window size can be changed with the win.size parameter of the function proc.chromR().

head(chrom@win.info)

# While loading entire genomes into memory may not be practical due to resource limitations, it is frequently practical to break a genome up into fractions that can be processed given the resources available on any system.
```

```{r, genetic_diff}
# A fundamental question to most population studies is whether populations are diverse and whether this diversity is shared among the populations? To address the question of within population diversity geneticists typically report heterozygosity.

data(vcfR_example)

pop <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))

#In vcfR, the function genetic_diff() was implemented to measure population diversity and differentiation.Because VCF data typically do not include population information we’ll have to supply it as a factor. 

myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')

# Heterozygosities are weighted by the number of alleles observed in each population. This was inspired by hierfstat::pairwise.fst() which uses the number of individuals observed in each population to weight the heterozygosities.

# By using the number of alleles observed instead of the number of individuals we remove an assumption about how many alleles each individual may contribute. That is, we should be able to accomodate samples of mixed ploidy.

knitr::kable(head(myDiff[,1:15]))

# Note that in some populations zero alleles were observed. Populations with zero alleles reported heterozygosities of ‘NaN’ because of this absence of data.

knitr::kable(head(myDiff[,16:19]))

# We now have information for each variant in the VCF data. Because this is typically a large quantity of information, we’ll want to summarize it. One way is to take averages of the data.

knitr::kable(round(colMeans(myDiff[,c(3:9,16,19)], na.rm = TRUE), digits = 3))
```

```{r, violin_plot}
# Another way to summarize data is to use violin plots.

library(reshape2)
library(ggplot2)

dpf <- melt(myDiff[,c(3:8,19)], varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)

p <- ggplot(dpf, aes(x=variable, y=Depth)) + geom_violin(fill="#2ca25f", adjust = 1.2)
p <- p + xlab("")
p <- p + ylab("")
p <- p + theme_bw()
p
```

## Exercises Part II
```{r, exercise2}
# 1) You actually have everything you need to make a Manhattan plot. Can you figure out how to plot G′ST (y-axis) by genomic position (POS)?

myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')

plot(getPOS(vcf), myDiff$Gprimest, pch = 20, col = "#1E90FF44", xlab = "", ylab = "", ylim = c(0, 1), xaxt = "n")
axis(side = 1, at = seq(0, 1e5, by = 1e4), labels = seq(0, 100, by = 10))
title(xlab='Genomic position (Kbp)')
title(ylab = expression(italic("G'"["ST"])))
# 2) This Manhatttan plot shouldlook a bit unusual. Can you think of anything that may be wrong with this analysis?

table(pop)
# The sample is so small

# 3) Can you figure out how to zoom in on a particular region of a chromosome in chromoqc()?

chromoqc(chrom, dp.alpha = 66, xlim = c(2e05, 4e05))

# 4) Can you use the function queryMETA() to look for other data in your file that may be of interest?

queryMETA(vcf)
```
